### DISCUSSION TITLE
ContextVar does not work with lifetime / startup event

### DISCUSSION THREAD

**Discussion by elbaro** [fastapi/fastapi#10575] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from contextlib import asynccontextmanager
from contextvars import ContextVar
from fastapi import FastAPI

var = ContextVar("var")


@asynccontextmanager
async def lifespan(app: FastAPI):
    var.set(200)
    print("=========== var is set ===============")
    yield
    print("=========== var is cleared ===========")

app = FastAPI(
    title="AAA",
    description="BBB",
    lifespan=lifespan,
)

@app.get("/")
async def index():
    return str(var.get())
```

### Description

The above code results in InternalServerError:
```LookupError: <ContextVar name='var' at 0x7fec896d3740>```

Extracting `var.set(..)` out of lifespan makes it working:

```py
var.set(200)

@asynccontextmanager
async def lifespan(app: FastAPI):
    print("=========== var is set ===============")
    yield
    print("=========== var is cleared ===========")
```
However I need an async initialization with `var.set(await async_init())`, and `lifespan` is the only place I can run async code.

How can you asynchronously initialize a global resource with ContextVar?

### Operating System

Linux

### FastAPI Version

0.104.1

### Pydantic Version

2.4.2

### Python Version

Python 3.11.4

**Comment by jgould22** [comment, marked as answer]
ContextVar is working as expected here I think.

If you set it in the lifespan function that is a different context, so it is not set in the task that is spawned for your request. Lifespan is run at the start of the application, not every request.

**Reply by elbaro** [reply]
Yes every reqeuest not inheriting the context is the problem. It should inherit.

What you say is that in the following code, `SomeContextManager` is not applied to every request. Then what's the point of using the context manager for lifespan?
```
@asynccontextmanager
async def lifespan(app: FastAPI):
   with SomeContextManager():
       yield
```

**Reply by ItachiSan** [reply]
Hi, I do disagree together with elbaro; in ASGI & asyncio context variables should allow us to share stuff between requests.
We are able to create stuff in the lifespan function but not "record" it for later usage; middlewares thus do things outside of the lifespan.
Both starlette-context and fastapy_async_sqlalchemy do set context variables as mentioned.
Given that lifespan is assumed to use a simple global variable (see the tutorial), I believe context variables should work the same.

I am thinking of opening an issue so that we can get the ball rolling, what do you think?

**Reply by Kludex** [reply]
@jgould22 's answer is correct. It's a different context, nothing to do about it. It's not an opinion, it's a fact, there's nothing to disagree about it.

You folks probably want to use [Lifespan State](https://www.starlette.io/lifespan/#lifespan-state).

**Comment by EdenHell** [comment]
I encountered the same issue, and I resolved it this way:

```python
# Maybe you can also use middleware
class APIRoute(fastapi.routing.APIRoute):
    def get_route_handler(self):
        original_route_handler = super().get_route_handler()

        async def custom_route_handler(request: Request):
            with ctx_vars(
                request.state.settings, request.state.db, request.state.executor
            ):
                return await original_route_handler(request)

        return custom_route_handler


@asynccontextmanager
async def lifespan(app: FastAPI):
    async with some_context_manager() as (settings, db, executor):
        yield {"settings": settings, "db": db, "executor": executor}


app = FastAPI(lifespan=lifespan)
app.router.route_class = APIRoute
```
